h/search/query.py

```diff
@@ -242,6 +275,91 @@ def __call__(self, search, params):
             'terms', **{'target.scope': list(uris)})
 
 
+class UriCombinedWildcardFilter(object):
+
+    """
+    A filter that selects only annotations where the uri matches.
+
+    If separate_keys is True:
+        Wildcard searches are only performed on wildcard_uri's, and exact match searches
+        are performed on uri/url parameters.
+    If separate_keys is False:
+        uri/url parameters are expected to contain exact matches and wildcard matches.
+        Values containing wildcards are intrepted as wildcard searches.
+
+    If specifying a uri with wildcards:
+    * will match any character sequence (including an empty one), and a ? will match
+    any single character.
+    """
+
+    def __init__(self, request, separate_keys=False):
+        """Initialize a new UriFilter.
+
+        :param request: the pyramid.request object
+        :param separate_keys: if True will treate wildcard_uri as wildcards and uri/url
+            as exact match. If False will treat any values in uri/url containing wildcards
+            ("?" or "*") as wildcard searches.
+
+        """
+        self.request = request
+        self.separate_keys = separate_keys
+
+    def __call__(self, search, params):
+        # If there are no uris to search for there is nothing to do so return early.
+        if not ("uri" in params or "url" in params or "wildcard_uri" in params):
+            return search
+
+        if self.separate_keys:
+            uris = _pop_param_values(params, 'uri') + _pop_param_values(params, 'url')
+            wildcard_uris = _pop_param_values(params, 'wildcard_uri')
+        else:
+            uris = _pop_param_values(params, 'uri') + _pop_param_values(params, 'url')
+            # Split into wildcard uris and non wildcard uris.
+            wildcard_uris = [u for u in uris if "*" in u or "?" in u]
+            uris = [u for u in uris if "*" not in u and "?" not in u]
+
+        # Ignore all invalid wildcard uris.
+        wildcard_uris = self._normalize_uris(
+            [u for u in wildcard_uris if wildcard_uri_is_valid(u)],
+            normalize_method=self._wildcard_uri_normalized)
+        uris = self._normalize_uris(uris)
+
+        queries = []
+        if wildcard_uris:
+            queries = [Q('wildcard', **{"target.scope": u}) for u in wildcard_uris]
+        if uris:
+            queries.append(Q("terms", **{'target.scope': uris}))
+        return search.query('bool', should=queries)
+
+    def _normalize_uris(self, query_uris, normalize_method=uri.normalize):
```

Comment:
In testing I discovered an unfortunate issue with our URI normalization. If the path is present but empty, it is removed from the normalized URI. This means that a wildcard search for `https://example.org/*` will not match an annotation on `https://example.org/` since the latter URL gets normalized to `httpx://example.org` in the search index.

I would suggest not trying to fix that in this PR but leaving it as follow-up work.

Reply:
^ I think this is the number 1 use case and I really rather address it right now.

Reply:
I agree it is a little unfortunate but my thinking there was this would be handled via user input aka: the user would search for https://example.org/* and https://example.org. The problem is when you want to match uri's like httpx://example.org?page=24 since you aren't allowed to specify *'s in the domain of the uri right now.

Reply:
@robertknight or @hmstepanek can we track this in another issue. Don't think its a huge issue if we don't fix it now, but we should definitely address this in the future. 
> his means that a wildcard search for https://example.org/* will not match an annotation on https://example.org/ since the latter URL gets normalized to httpx://example.org in the search index.
I would suggest not trying to fix that in this PR but leaving it as follow-up work.

Reply:
I have fixed this but as a consequence it excepts `?` inside the domain (not in the scheme just the netloc). I figure this is ok for now and if it causes slowdown issues we can fix it. I don't want to merge this without full-filing the basic requirements and I feel like this is a basic requirement. You should be able to wildcard search for  https://example.org*

Reply:
I had a better idea for implementation this morning so current state is: we still don't allow wildcard inside the http://foo.com/ but http://foo.com/* normalizes to http://foo.com* now so it does include _all_ domain uri results now.